### DISCUSSION TITLE
Parametrise custom gate

### DISCUSSION THREAD

**Discussion by MirandaCarou** [tiago939/QuForge#5] [Q&A] [answered]
Hi! I was wondering if it is possible to create a parameterised custom gate. As far as I know, it is possible to implement a custom unitary gate, but after checking the documentation, I don’t see any parameter that would allow me to include, for example, an angle as in a rotation gate. Is there any way to do this, or is it not implemented yet?

Thanks in advance!

**Comment by tiago939** [comment]
Hi Miranda!
There a few different ways you could implement a custom gate.

If you have a known unitary matrix, you could implement it using the custom U gate:
```
import quforge.quforge as qf
import torch
import numpy as np

a, b = np.cos(0.55), np.sin(0.55)
matrix = torch.tensor([
    [a, -1j*b],
    [-1j*b, a]
], dtype=torch.complex64)

gate = qf.U(matrix=matrix, dim=2, wires=1)
state = qf.State('0', dim=2)
result = gate(state)
```

If your specific parametrisation requires a rotation gate, you can feed the angle directly as an input parameter during gate/circuit initialization:
```
import quforge.quforge as qf

initial_angle = 0.55
gate = qf.RX(angle=initial_angle)

state = qf.State('0')
result = gate(state)
```

or during the forward process:
```
import quforge.quforge as qf

your_angle = 0.55
gate = qf.RX()
state = qf.State('0')

result = gate(state, param=[your_angle])
```

Just be aware that depending on your objective, this angle or parametrization can be differentiable or not.

**Reply by MirandaCarou** [reply]
Hi again,

Thanks for your fast answer! In my case, I’m using the first option, so most of the gates in my model are custom gates, and I need them to be parametrised, including the rotation gates (because I’m working with a different representation of qutrits, called the Majorana stellar representation). At the same time, these gates are part of my QML model, so I need these parameters to be differentiable. Does this mean that since all my gates are unitary, parametrised custom gates, they cannot be differentiable?

**Reply by tiago939** [reply, marked as answer]
Hi Miranda!

By default, all rotation gates are parametrized. Unless you use with `torch.no_grad()`, they will have a derivative when you call `loss.backward()`.

In case of a custom parametrized gate, you can tell the optimizer that your custom matrix is parameterized. Something like this:
```
import quforge.quforge as qf
import torch
import numpy as np

a, b = np.cos(0.55), np.sin(0.55)
matrix = torch.nn.Parameter(torch.tensor([
    [a, -1j*b],
    [-1j*b, a]
], dtype=torch.complex64))

input_state = qf.State('0', dim=2)
optimizer = torch.optim.Adam([matrix], lr=1e-3)

custom_gate = qf.U(dim=2, wires=1)
custom_gate.M = matrix
custom_gate.random = False

for epoch in range(100):
    optimizer.zero_grad()
    output_state = custom_gate(input_state)
    
    loss = torch.mean(torch.abs(input_state - output_state))
    loss.backward()
    optimizer.step()
```

If you want to include it inside a circuit, you can do something like this:
```
import quforge.quforge as qf
import torch
import numpy as np

a, b = np.cos(0.55), np.sin(0.55)
matrix = torch.nn.Parameter(torch.tensor([
    [a, -1j*b],
    [-1j*b, a]
], dtype=torch.complex64))

circuit = qf.Circuit(dim=2, wires=2)
circuit.H(index=[0,1])
circuit.RX(index=[0])
circuit.RY(index=[1])
circuit.U(index=[0])

circuit.circuit[3].M = matrix
circuit.circuit[3].random = False

input_state = qf.State('0-0', dim=2)
optimizer = torch.optim.Adam(circuit.parameters(), lr=1e-3)

for epoch in range(100):
    optimizer.zero_grad()
    output_state = circuit(input_state)

    loss = torch.mean(torch.abs(input_state - output_state))
    loss.backward()
    optimizer.step()
```

Please let me know if this helps.

**Reply by MirandaCarou** [reply]
Thanks, it helps! :)
